src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -180,62 +178,102 @@ private abstract class ReadOperation : AsyncOperation
 
         private abstract class WriteOperation : AsyncOperation 
         {
-        }        
+        }
 
-        private sealed class SendOperation : WriteOperation
+        private abstract class SendOperation : WriteOperation
         {
-            public byte[] Buffer;
-            public int Offset;
-            public int Count;
             public SocketFlags Flags;
             public int BytesTransferred;
-            public IList<ArraySegment<byte>> Buffers;
-            public int BufferIndex;
+            public int Offset;
+            public int Count;
 
             protected sealed override void Abort() { }
 
             public Action<int, byte[], int, SocketFlags, SocketError> Callback
             {
-                private get { return (Action<int, byte[], int, SocketFlags, SocketError>)CallbackOrEvent; }
-                set { CallbackOrEvent = value; }
+                set => CallbackOrEvent = value;
             }
 
-            protected sealed override void InvokeCallback()
+            protected sealed override void InvokeCallback() =>
+                ((Action<int, byte[], int, SocketFlags, SocketError>)CallbackOrEvent)(BytesTransferred, SocketAddress, SocketAddressLen, SocketFlags.None, ErrorCode);
+        }
+
+        private sealed class BufferArraySendOperation : SendOperation
```

Comment:
We could, but then we'd need to pin the buffer for potentially much longer than we otherwise need to.  The byte[] only needs to be pinned during the synchronous call to send, but to use the BufferPtr version, we need to pin it the whole time in order to get the pointer and keep it valid.